uamqp/management_operation.py

```diff
@@ -45,37 +46,63 @@ def _on_amqp_management_open_complete(self, result):
     def _on_amqp_management_error(self):
         """Callback run if an error occurs in the send/receive links."""
         # TODO: This probably shouldn't be ValueError
-        self.mgmt_error = ValueError("Management Operation error ocurred.")
+        self.mgmt_error = ValueError("Management Operation error occurred.")
+
+    def _on_execute_operation_complete(
+        self,
+        operation_id,
+        operation_result,
+        status_code,
+        status_description,
+        raw_message,
+        error=None
+    ):
+        _LOGGER.debug(
+            "mgmt operation completed, operation id: %r; operation_result: %r; status_code: %r; "
+            "status_description: %r, raw_message: %r, error: %r",
+            operation_id,
+            operation_result,
+            status_code,
+            status_description,
+            raw_message,
+            error
+        )
+
+        if operation_result in\
+                (ManagementExecuteOperationResult.ERROR, ManagementExecuteOperationResult.INSTANCE_CLOSED):
+            self.mgmt_error = error
+            _LOGGER.error(
+                "Failed to complete mgmt operation due to error: %r. The management request message is: %r",
+                error, raw_message
+            )
+        else:
+            self._responses[operation_id] = (status_code, status_description, raw_message)
 
     def execute(self, message, operation=None, operation_type=None, timeout=0):
         start_time = time.time()
         operation_id = str(uuid.uuid4())
         self._responses[operation_id] = None
-
-        def on_complete(operation_result, status_code, status_description, raw_message, error_condition):
-            if operation_result != ManagementExecuteOperationResult.OK:
-                _LOGGER.error(
-                    "Failed to complete mgmt operation.\nStatus code: %r\nMessage: %r",
-                    status_code, status_description)
-            #message = Message(message=wrapped_message) if wrapped_message else None
-            self._responses[operation_id] = (status_code, raw_message, status_description)
+        self.mgmt_error = None
 
         self._mgmt_link.execute_operation(
```

Comment:
So `execute_operation` returns once the request has been sent, then the while loop below is used to wait for the response, is that correct?

Reply:
yep, your understanding is correct